### PR TITLE
Feature/api battle create team

### PR DESCRIPTION
**Pull request by DeborahAlmeida**
## Objective and Motivation
- Make endpoint to create teams

### CODE REVIEW

File: backend/services/create_team.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+from battle.battles.battle import validate_sum_pokemons, verify_pokemon_is_saved
+
+from pokemon.helpers import verify_pokemon_exists_api
+
+
+def verify_if_data_is_valid(data):  # pylint: disable=too-many-return-statements
```
Comment (rsarai) [inline backend/services/create_team.py]: This is already an improvement on code duplication, however, this is still not optimal.
1. First, this function makes multiple validations, this is usually not a problem, but in this case, the name is too generic, by looking at the serializers and forms I don't know what is being validated. I advise you to split each validation into an individual function and calling them on the forms and serializers.
2. This function has multiple return types (boolean and strings), I strongly advise you to avoid this. This makes the behavior obscure, remember explicit is better than implicit.

File: backend/services/create_team.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    message_error = None
+
+    if 'battle' not in data:
+        message_error = 'ERROR: Select a valid battle'
+        return message_error
```
Comment (rsarai) [inline backend/services/create_team.py]: This may be required on the form, but not on the serializer since you can set a field to be required

File: backend/services/create_team.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    if data['trainer'] not in (data['battle'].creator, data['battle'].opponent):
+        message_error = ("ERROR: You do not have permission for this action.")
+        return message_error
```
Comment (rsarai) [inline backend/services/create_team.py]: This validation on DRF should be a permission:  https://www.django-rest-framework.org/api-guide/permissions/#custom-permissions

Comment (rsarai) [reply backend/services/create_team.py]: For the form, this should be a mixin, you can use the LoginRequired mixin as example: https://docs.djangoproject.com/en/3.2/topics/auth/default/#the-loginrequired-mixin

File: backend/services/create_team.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        message_error = 'ERROR: Pokemons sum more than 600 points. Select again'
+        return message_error
+
+    if data['position_pkn_1'] in (data['position_pkn_2'], data['position_pkn_3']):
+        message_error = 'ERROR: You cannot add the same position'
+        return message_error
+
+    if data['position_pkn_2'] == data['position_pkn_3']:
+        message_error = 'ERROR: You cannot add the same position'
+        return message_error
```
Comment (rsarai) [inline backend/services/create_team.py]: This validation can be improved. You can do this using sets: 
```
set(data['position_pkn_1'], data['position_pkn_2'], data['position_pkn_3']) != set(1,2,3)
```

File: backend/services/create_team.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    verify_pokemon_is_saved(
```
Comment (rsarai) [inline backend/services/create_team.py]: This name is still misleading, isn't this saving the pokemon on the database if they not exist?

File: backend/services/create_team.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+from battle.battles.battle import validate_sum_pokemons, verify_pokemon_is_saved
```
Comment (rsarai) [inline backend/services/create_team.py]: The name of this folder is confusing. My advice for naming services is to use the name of the app they are in. In this case, `services/battle.py`

File: backend/battle/serializers.py [hunk 7]
```diff
@@ -19,6 +28,98 @@ class Meta:
...
+        position_pkn_1 = self.validated_data['position_pkn_1']
+        position_pkn_2 = self.validated_data['position_pkn_2']
+        position_pkn_3 = self.validated_data['position_pkn_3']
+
+        validated_data.pop('pokemon_1')
+        validated_data.pop('pokemon_2')
+        validated_data.pop('pokemon_3')
+        validated_data.pop('position_pkn_1')
+        validated_data.pop('position_pkn_2')
+        validated_data.pop('position_pkn_3')
```
Comment (rsarai) [inline backend/battle/serializers.py]: hmmm, any specific reason for adding this?

File: backend/battle/serializers.py [hunk 8]
```diff
@@ -19,6 +28,98 @@ class Meta:
...
+            "pokemon_3"
+        )
+
+    def validate(self, attrs):
+        valid = verify_if_data_is_valid(attrs)
+        if valid is not True:
+            raise serializers.ValidationError(valid)
+        return attrs
+
+    def create(self, validated_data):
```
Comment (rsarai) [inline backend/battle/serializers.py:109]: I wonder how are you testing that this is working?

**Review by rsarai — CHANGES_REQUESTED** [review]

File: backend/battle/serializers.py [hunk 9]
```diff
@@ -19,6 +28,98 @@ class Meta:
...
+            team=instance,
+            pokemon=pokemon_3_object,
+            order=position_pkn_3)
+        try:
+            Team.objects.get(
+                trainer=validated_data['battle'].creator,
+                battle=validated_data['battle'])
+            Team.objects.get(
+                trainer=validated_data['battle'].opponent,
+                battle=validated_data['battle'])
```
Comment (rsarai) [inline backend/battle/serializers.py]: hmmmm, why is this required?

File: backend/services/create_team.py (hunk 2 above)
Comment (DeborahAlmeida) [reply backend/services/create_team.py]: hmm great!

File: backend/services/create_team.py (hunk 5 above)
Comment (DeborahAlmeida) [reply backend/services/create_team.py]: yes, I will improve this name function :D

File: backend/services/create_team.py (hunk 6 above)
Comment (DeborahAlmeida) [reply backend/services/create_team.py]: thanks :)

File: backend/battle/serializers.py (hunk 8 above)
Comment (DeborahAlmeida) [reply backend/battle/serializers.py:109]: What do you mean?

File: backend/battle/serializers.py (hunk 9 above)
Comment (DeborahAlmeida) [reply backend/battle/serializers.py]: I verify if exists the creator and opponent's teams

File: backend/services/create_team.py (hunk 3 above)
Comment (DeborahAlmeida) [reply backend/services/create_team.py]: thanks for all the support :dancers:

File: backend/services/create_team.py (hunk 4 above)
Comment (DeborahAlmeida) [reply backend/services/create_team.py]: thank you :D

File: backend/battle/battles/battle.py [hunk 10]
```diff
@@ -53,7 +53,7 @@ def validate_sum_pokemons(pokemons):
     return total_points <= 600
 
 
-def verify_pokemon_is_saved(pokemons):
+def save_pokemons_if_they_doenst_exist(pokemons):
```
Comment (rsarai) [inline backend/battle/battles/battle.py]: Typo: `save_pokemons_if_they_dont_exist`

File: backend/battle/forms.py [hunk 11]
```diff
@@ -2,17 +2,22 @@
 from django.contrib.auth.forms import UserCreationForm
 
 from battle.models import Battle, PokemonTeam, Team
-from battle.battles.battle import validate_sum_pokemons, verify_pokemon_is_saved
+from battle.battles.battle import save_pokemons_if_they_doenst_exist
```
Comment (rsarai) [inline backend/battle/forms.py]: This namespace is weird: `battle.battles.battle`. I'm not gonna stress about this but we should avoid this going forward

File: backend/battle/forms.py [hunk 12]
```diff
@@ -2,17 +2,22 @@
...
+from services.battle import (
     validate_if_creator_and_opponent_has_different_contenders,
-    fetch_opponent_or_create_if_doenst_exist, send_invite_email_or_create_password_email)
+    fetch_opponent_or_create_if_doenst_exist,
+    send_invite_email_or_create_password_email,
+    verify_pokemons_fields_has_missing_values,
+    verify_fields_has_wrong_pokemon_name,
+    verify_positions_fields_has_missing_values,
+    verify_pokemon_sum_is_valid,
+    verify_positions_fields_has_duplicate_values,
```
Comment (rsarai) [inline backend/battle/forms.py]: I think all these helpers that start with `verify` should start with `has` or `is`

File: backend/battle/serializers.py [hunk 13]
```diff
@@ -19,6 +35,109 @@ class Meta:
...
+        label="Position",
+        required=False)
+    position_pkn_2 = serializers.ChoiceField(
+        choices=POSITION_CHOICES,
+        label="Position",
+        required=False)
+    position_pkn_3 = serializers.ChoiceField(
+        choices=POSITION_CHOICES,
+        label="Position",
+        required=False)
```
Comment (rsarai) [inline backend/battle/serializers.py:58]: These fields should be required no?

File: backend/battle/serializers.py [hunk 14]
```diff
@@ -19,6 +35,109 @@ class Meta:
...
+
+        valid = verify_positions_fields_has_duplicate_values(attrs)
+        if valid is not True:
+            raise serializers.ValidationError(valid)
+
+        save_pokemons_if_they_doenst_exist(
+            [
+                attrs['pokemon_1'],
+                attrs['pokemon_2'],
+                attrs['pokemon_3']])
```
Comment (rsarai) [inline backend/battle/serializers.py]: Indentation is a bit weird here.

File: backend/battle/serializers.py [hunk 15]
```diff
@@ -19,6 +35,109 @@ class Meta:
...
+            team=instance,
+            pokemon=pokemon_2_object,
+            order=position_pkn_2)
+
+        PokemonTeam.objects.create(
+            team=instance,
+            pokemon=pokemon_3_object,
+            order=position_pkn_3)
+
+        if len(list(validated_data['battle'].teams.all())) == 2:
```
Comment (rsarai) [inline backend/battle/serializers.py]: Do you need the `list` here?

File: backend/services/battle.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+            email_template_name='registration/guest_email.html',
+            from_email=settings.FROM_EMAIL,)
+    else:
+        send_invite_email(opponent, creator)
+
+
+def verify_pokemons_fields_has_missing_values(data):
+    if ('pokemon_1' or 'pokemon_2' or 'pokemon_3') not in data:
+        message_error = 'ERROR: Select all pokemons'
+        return message_error
```
Comment (rsarai) [inline backend/services/battle.py]: As I mentioned before, please avoid creating functions that return missed types, in this case, boolean and string. The error messages should be in the form and in the serializer

File: backend/services/battle.py [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    return True
+
+
+def verify_positions_fields_has_missing_values(data):
+    if ('position_pkn_1' or 'position_pkn_2' or 'position_pkn_3') not in data:
+        message_error = 'ERROR: Select all positions'
+        return message_error
```
Comment (rsarai) [inline backend/services/battle.py]: Same as before

File: backend/services/battle.py [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    return True
+
+
+def verify_fields_has_wrong_pokemon_name(data):
+    pokemons_exist = verify_pokemon_exists_api(
+        [data['pokemon_1'], data['pokemon_2'], data['pokemon_3']])
+
+    if not pokemons_exist:
+        message_error = 'ERROR: Type the correct pokemons name'
+        return message_error
```
Comment (rsarai) [inline backend/services/battle.py]: Same as before

File: backend/services/battle.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    return True
+
+
+def verify_pokemon_sum_is_valid(data):
+    valid_pokemons = validate_sum_pokemons(
+        [data['pokemon_1'], data['pokemon_2'], data['pokemon_3']])
+
+    if not valid_pokemons:
+        message_error = 'ERROR: Pokemons sum more than 600 points. Select again'
+        return message_error
```
Comment (rsarai) [inline backend/services/battle.py]: Same as before

File: backend/services/battle.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def verify_positions_fields_has_duplicate_values(data):
+    positions = {
+        int(data['position_pkn_1']),
+        int(data['position_pkn_2']),
+        int(data['position_pkn_3'])}
+    if positions != {1, 2, 3}:
+        message_error = 'ERROR: You cannot add the same position'
+        return message_error
```
Comment (rsarai) [inline backend/services/battle.py]: Same as before

File: backend/services/battle.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
+from users.models import User
```
Comment (rsarai) [inline backend/services/battle.py:1]: As I mentioned before all helpers that start with `verify` should start with `has` or `is`

**Review by rsarai — CHANGES_REQUESTED** [review]

File: backend/battle/forms.py (hunk 11 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py]: okay, thanks for alert me

File: backend/battle/forms.py (hunk 12 above)
Comment (DeborahAlmeida) [reply backend/battle/forms.py]: hmm okay! I misunderstood, thanks!

File: backend/battle/serializers.py (hunk 13 above)
Comment (DeborahAlmeida) [reply backend/battle/serializers.py:58]: true!

Comment (DeborahAlmeida) [reply backend/battle/serializers.py:58]: Actually, if I set true, it returns an error, so I need to set False, and I verify if the field is empty on serializer

File: backend/battle/serializers.py (hunk 14 above)
Comment (DeborahAlmeida) [reply backend/battle/serializers.py]: Okay, is this final, right?

File: backend/battle/serializers.py (hunk 15 above)
Comment (DeborahAlmeida) [reply backend/battle/serializers.py]: actually, I don't, thanks!

File: backend/services/battle.py (hunk 16 above)
Comment (DeborahAlmeida) [reply backend/services/battle.py]: hmm interesting, thanks!

File: backend/services/battle.py (hunk 21 above)
Comment (DeborahAlmeida) [reply backend/services/battle.py:1]: thank you :)
